test/gradcheck.jl

```diff
@@ -2,28 +2,52 @@ using Zygote, NNlib, Test, Random, LinearAlgebra, Statistics, FillArrays, FFTW
 using Zygote: gradient
 using NNlib: conv, ∇conv_data, depthwiseconv
 using Base.Broadcast: broadcast_shape
+using FiniteDifferences: FiniteDifferences, central_fdm, forward_fdm, backward_fdm
 
-function ngradient(f, xs::AbstractArray...)
-  grads = zero.(xs)
-  for (x, Δ) in zip(xs, grads), i in 1:length(x)
-    δ = sqrt(eps())
-    tmp = x[i]
-    x[i] = tmp - δ/2
-    y1 = f(xs...)
-    x[i] = tmp + δ/2
-    y2 = f(xs...)
-    x[i] = tmp
-    Δ[i] = (y2-y1)/δ
+realdomainrange(::Any) = (-Inf, Inf)
+realdomainrange(::Union{typeof.((acos,asin,atanh))...}) = (-1, 1)
+realdomainrange(::typeof(acosh)) = (1, Inf)
+realdomainrange(::Union{typeof.((log,sqrt,^))...}) = (0, Inf)
+
+function default_fdm(f::F) where F
+  # Attempt to choose a way of finite differencing that will avoid escaping the domain
+  lower, upper = realdomainrange(F)
+  if lower == -Inf && upper==Inf  # Ideal case
+    central_fdm(5, 1)
+  elseif upper == Inf
+    forward_fdm(5, 1)
+  elseif lower == -Inf
+    backward_fdm(5,1)
+  else  # fallback, hopefully input is not near bounds
+    central_fdm(5, 1)
+  end
+end
+
+function ngradient(f, xs::AbstractArray...; fdm=default_fdm(f))
+    println(" ")  # make sure TravisCI doesn't time out
+    return FiniteDifferences.grad(fdm, f, xs...)
+end
+
+function gradcheck(f, xs...; kwargs...)
+  fin_grads = ngradient(f, xs...; kwargs...)
```

Comment:
does this work for Array-based functions? and how do we integrate this with #235 I think the reason why I'm using j'vp in that PR is because this provide a more straightforward way to test these kind of functions.

Reply:
Oh it uses `sum` in `gradtest`, nvm

Reply:
Yeah. That also should change, but one PR at a time.